### PR TITLE
[WIP] CouchDB support

### PR DESCRIPTION
**Pull request by gregkare**

### CODE REVIEW

File: lib/remote_storage/couchdb.rb [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+    def do_put_request(url, data, content_type)
+      begin
+        res = RestClient.get(url)
+        rev = JSON.parse(res.body)["_rev"]
+        url = "#{url}?rev=#{rev}"
+      rescue RestClient::ResourceNotFound
+        #do nothing
+      end
+      mime_type = MIME::Types[content_type].first
+      if mime_type.content_type == "application/json" || !mime_type.binary?
```
Comment (galfert) [inline lib/remote_storage/couchdb.rb]: When `mime_type.content_type` is "application/json", then `mime_type.binary?` should also be `false`, no?

Then `if !mime_type.binary?` should be enought here.

Comment (skddc) [inline lib/remote_storage/couchdb.rb]: Also see my comment above.

File: config.yml.example [hunk 2]
```diff
@@ -16,6 +16,8 @@ development: &defaults
   # # uncomment this section and comment the riak one
   # swift: &swift_defaults
   #   host: "https://swift.example.com"
+  # couchdb: #TODO create the database using chef
+  #   uri: "http://localhost:5984/enterprise_storage"
```
Comment (skddc) [inline config.yml.example]: This would actually have to be a database per Managed Storage customer.

File: lib/remote_storage/couchdb.rb (hunk 1 above)
Comment (gregkare) [inline lib/remote_storage/couchdb.rb]: application/json is actually a binary type, we didn't catch it on Wednesday

**Comment by skddc** [comment]
I'm not sure how the latest commit makes sense. The whole point of us evaluating CouchDB for Storage is so that one can do map/reduce queries on the JSON data. For that, it can't be an attachment. Did I miss something here?

**Comment by gregkare** [comment]
@skddc Shit, I forgot about map/reduce. I'm now trying to fix the UTF-8 bug without storing everything as an attachment

**Comment by gregkare** [comment]
I found another bug with how we handle JSON right now (we're returning the JSON parsed as Ruby), writing a spec for it

**Comment by gregkare** [comment]
ASCII-8BIT is binary. Adding `.force_encoding('UTF-8')` to the data works. We used `force_encoding` in liquor-cabinet's riak provider (https://github.com/5apps/liquor-cabinet/blob/e994696fe03dc001e392b396aeb61a60f4341bac/lib/remote_storage/riak.rb#L423-L432), but only to check if the data is binary

**Comment by gregkare** [comment]
I've pushed my fix. Specs are still green (I added two new ones) as well as the API test suite

**Comment by skddc** [comment]
This doesn't solve the root issue. WHY is JSON data coming in as binary?!

**Comment by gregkare** [comment]
It's not just JSON, all PUT data in Sinatra is read as binary (it's a StringIO instance encoded in ASCII-8BIT). It's not an issue in the other providers because we don't save this data in a JSON object (it's not actually binary, it's just that Ruby doesn't know. It could be ASCII or UTF-8 or something else)

I think that forcing the encoding is the right way to go in this case because JSON is supposed to be Unicode

**Comment by galfert** [comment]
FYI: `remoteStorage.js` is also adding a `charset=UTF-8` and `charset=binary` to the `content-type` header on PUTs.

**Comment by gregkare** [comment]
The way we stored the JSON data also doesn't allow to do map/reduce on it. I'm changing it now

**Comment by skddc** [comment]
What's the status here?

**Comment by gregkare** [comment]
My last commit from June allows to run map/reduce on JSON documents. However, when doing GET requests I'm removing the built-in keys that CouchDB adds to documents, for example `_rev` and `_id`, by parsing the JSON, removing keys and returning a JSON represention. The current status is that two specs from the API test suite are failing because of formatting of the returned JSON docs: the ones I'm returning do not have spaces between the keys and values because of the way the Ruby JSON module works

**Comment by skddc** [comment]
1. How is map/reduce relevant for just the object storage backend here?
2. Should we care about the format being precisely the same at this point in time?

**Comment by galfert** [comment]
> Should we care about the format being precisely the same at this point in time?

Sounds to me like it's rather the API test suite that needs fixing. As long as it's valid JSON and all expected content is there (and nothing else), it shouldn't care for the formatting.

**Comment by skddc** [comment]
I disagree for general usage by end users. We actually had the same issue with Riak in the past, and had users complain about it. It's just not polite to transform data from end users and returning something that they didn't store (no matter if the changed bytes are whitespace and line breaks or not).

I was asking, because we intended this to be used where it really doesn't matter, as users don't store their own data.

**Comment by galfert** [comment]
Oh, right. I don't know why, but I only thought about the directory listing JSON, not users' own JSON.

**Pull request closed without merging**